### DISCUSSION TITLE
How to mark a texture as used / handling json pointers

### DISCUSSION THREAD

**Discussion by marwie** [donmccurdy/glTF-Transform#696] [Q&A] [answered]
Hey,

we use glTF transform to pass through our extensions during compression at [Needle](https://needle.tools) and I'm currently in the process of upgrading to glTF transform 2.x.

We support / work with json pointer extensions to e.g. resolve references to textures or materials. 
For example a custom component that declared in our components extensions has a field ``myTexture`` with a reference to a texture as ``/textures/0``. 

Here are some things that I noticed if the texture is not used by any material:
1) the ``json.textures`` array will not contain the texture definition entry after compression
2) the ``json.images`` array still has the texture image data (so the buffer is still in the glb)
3) the ``json.samplers`` array also does not contain the sampler anymore that was referenced by the texture

I think 2) might be a bug since when the texture and sampler will be thrown I think out the content of the buffer should also be discarded (since it is not used by anyone)

About (1) and (3) I wanted to aks how to properly mark a resource (a texture in this case) as used so it won - I read I can use the property graph for this, maybe you can give me a hint on how to use it or where to find more information? 

For now I have a workaround that uses the ``imageIndexMap`` to find the new index of the previously referenced image and then checks if the texture is still in ``textures`` and if not it re-inserts it and also re-creates the missing sampler (if it is missing) - but this feels quite hacky :)

Attached is a simple example here: [cube.zip](https://github.com/donmccurdy/glTF-Transform/files/9692524/cube.zip)
and for comparison here is the sample cube processed with the workaround in place: [cube.needle_opaque.glb.zip](https://github.com/donmccurdy/glTF-Transform/files/9692552/cube.needle_opaque.glb.zip)

---
Before transform:
![image](https://user-images.githubusercontent.com/5083203/193455290-f4d0bad4-bc1b-4e80-83e7-b6ae79357c4e.png)

After transform:
![image](https://user-images.githubusercontent.com/5083203/193455330-a86aac82-4336-4567-91b9-d09c5a2bf82c.png)

**Comment by donmccurdy** [comment, marked as answer]
Hi @marwie! The way that "unused" resources are written, or not written, in cases (1), (2), and (3) goes back to [conceptual simplifications](https://gltf-transform.donmccurdy.com/concepts.html) in glTF-Transform, compared to the glTF specification.  glTF Transform does not directly represent glTF "json.textures", "json.samplers", or "json.bufferViews" objects in its internal structure, instead it figures out what they must be (based on references among other resources) and generates those definitions at export time. This makes processing considerably simpler, without sacrificing any expressivity that — IMHO – matters.

Of the resources glTF _does_ directly represent, it does not omit things from the output just because they are unused. This means that loose images, materials, meshes, accessors, etc. can be written if you wish. The `prune` function is there to remove them, if not.

**tl;dr** – (1), (2), and (3) are intentional, although as you've seen, for practical purposes you'll usually want to make sure glTF-Transform knows how resources are being used. There are other reasons to do so; the reference to a [Texture](https://gltf-transform.donmccurdy.com/classes/core.texture.html) also provides context needed to choose the right compression settings, or (possible future work) to modify UV maps. Hopefully that's acceptable. For the rest of this response I'll focus on how to indicate that a texture or other resource is being used.

***

The preferred way to implement an extension in v2 would be this —

1. During import, convert pointers (whether index-based or JSON pointers) to graph edges
2. During export, write the resources first and then convert back to appropriate serialized pointers

Every [Property](https://gltf-transform.donmccurdy.com/classes/core.property.html) or [ExtensionProperty](https://gltf-transform.donmccurdy.com/classes/core.extensionproperty.html) is an instance of a [`property-graph` GraphNode](https://github.com/donmccurdy/property-graph#api). An extension uses the GraphNode API internally to implement any public getters/setters that it chooses to expose. 

The Transmission property ([`transmission.ts`](https://github.com/donmccurdy/glTF-Transform/blob/8dd6018/packages/extensions/src/khr-materials-transmission/transmission.ts#L29)) is a good example here. The texture is connected to the Transmission extension in the graph by the `setRef` call here—

https://github.com/donmccurdy/glTF-Transform/blob/8dd601845270686279dbf0b79f4e646bf6b89394/packages/extensions/src/khr-materials-transmission/transmission.ts#L77-L80

The channel metadata attached to that graph link is important, allowing KTX2 compression to make the right choices without knowing about all past/present/future material extensions. Calling `setRef` with a new value replaces the previous link, and passing `null` breaks the link entirely.

I find TypeScript helpful when implementing extensions, but plain JavaScript is fine too. The example at https://gltf-transform.donmccurdy.com/extensions.html#custom-extensions shows a simplified version of a similar extension, without TypeScript syntax.
